backend/src/ml_space_lambda/dataset/lambda_functions.py

```diff
@@ -258,35 +258,85 @@ def list_locations(event, context):
 
 @api_wrapper
 def list_files(event, context):
-    prefix = get_dataset_prefix(
+    env_variables = get_environment_variables()
+    query_string_parameters = event["queryStringParameters"] or {}
+    
+    # map query parameters keys to api parameter names
+    query_string_parameters = map_dict_keys(query_string_parameters, {
+        "nextToken": "ContinuationToken",
+        "pageSize": "MaxKeys",
+        "prefix": "Prefix"
+    })
+    
+    dataset_prefix = get_dataset_prefix(
         event["pathParameters"]["scope"], event["pathParameters"]["datasetName"]
     )
-
-    env_variables = get_environment_variables()
-    keys = []
-
-    if event["queryStringParameters"]:
-        s3_response = s3.list_objects_v2(
-            Bucket=env_variables["DATA_BUCKET"],
-            Prefix=prefix,
-            ContinuationToken=event["queryStringParameters"].get("nextToken", ""),
-        )
-    else:
-        s3_response = s3.list_objects_v2(Bucket=env_variables["DATA_BUCKET"], Prefix=prefix)
-
-    for object in s3_response.get("Contents", []):
-        keys.append({"key": object["Key"], "size": object["Size"]})
-
-    # Note: For s3_response if there are no files, the path to the dataset is returned
-    # Remove the dataset path to correctly reflect there are no files in the dataset
-    if len(keys) == 1 and keys[0]["key"] == prefix:
-        keys = []
-
-    response = {
-        "Keys": keys,
+    computed_prefix = "".join([dataset_prefix, query_string_parameters.get("Prefix", "")])
+    
+    query_parameters = {
+        "Bucket": env_variables["DATA_BUCKET"],
+        "Prefix": computed_prefix,
+        "Delimiter": "/"
     }
-
-    if "NextContinuationToken" in s3_response:
-        response["nextToken"] = s3_response["NextContinuationToken"]
-
-    return response
+    
+    # constrain which query parameters can be supplied
+    allowed_query_string_parameters = ["MaxKeys", "ContinuationToken"]
+    for key in filter_dict(query_string_parameters, allowed_query_string_parameters):
+        query_parameters[key] = query_string_parameters[key]
+
+    s3_response = s3.list_objects_v2(**query_parameters)
+
+    # drop unwated fields from response
+    preserved_response_keys = ["IsTruncated", "MaxKeys", "Prefix", "NextContinuationToken"]
+    response = filter_dict(s3_response, preserved_response_keys)
+    
+    response["Contents"] = []
+    
+    # drop unwated fields from contents
+    preserved_content_keys = ["Key", "Size"]
+    if "Contents" in s3_response:
+        response["Contents"].append(list(map(lambda object: filter_dict(object, preserved_content_keys), s3_response["Contents"])))
+    
+    # merge common prefixes with contents
+    if "CommonPrefixes" in s3_response:
+        response["Contents"].append(s3_response["CommonPrefixes"])
+    
+    # map response keys to api parameter names
+    return map_dict_keys(response, {
```

Comment:
Since we built the response object in this method do we really need to remap things at this point? I'm not against what's going on here just seems like a lot of indirection that makes things a little harder to follow.

Reply:
I had already used it earlier to map from `pageSize` -> `MaxKeys` and `nextToken` -> `ContinuationToken` at the begining so I thought it made sense to reuse it here. But if you think this adds too much indirection I can just replace these with assignments.